feat: diffusion training

ssmmnn11 opened the pull request:
## Description
Implements Diffusion training (EDM)

Full state prediction
Tendency prediction

todo: 
- [x] finalise main compatibility (e.g. keep batch sharded)
- [x] make sampler configurable in inference
- [x] clean up noise conditioning
- [x] tests are not completed / not all sensible
- [x] pydantic needs a sanity check
- [x] documentation

test of all configurations
- [x] deterministic
- [x] deterministic, bs > 1
- [ ] deterministic, sharded
- [x] ensemble
- [x] ensemble, bs > 1
- [ ] ensemble, sharded, members
- [ ] ensemble, sharded, members, model
- [x] diffusion tendency
- [x] diffusion tendency, bs > 1
- [x] diffusion tendency, sharded


<!-- readthedocs-preview anemoi-training start -->
----
📚 Documentation preview 📚: https://anemoi-training--401.org.readthedocs.build/en/401/

<!-- readthedocs-preview anemoi-training end -->

<!-- readthedocs-preview anemoi-graphs start -->
----
📚 Documentation preview 📚: https://anemoi-graphs--401.org.readthedocs.build/en/401/

<!-- readthedocs-preview anemoi-graphs end -->

<!-- readthedocs-preview anemoi-models start -->
----
📚 Documentation preview 📚: https://anemoi-models--401.org.readthedocs.build/en/401/

<!-- readthedocs-preview anemoi-models end -->

radiradev commented:
Very cool! Shouldn't samplers become part of `anemoi-inference` though?

radiradev commented:
I wouldn't be in favour of integrating `diffusers`, it would introduce a ton of extra dependencies. I quite like this lightweight library [az
ula](https://github.com/probabilists/azula)

ssmmnn11 commented:
> I wouldn't be in favour of integrating `diffusers`, it would introduce a ton of extra dependencies. I quite like this lightweight library [azula](https://github.com/probabilists/azula)

Can be an optional dependency.

gmertes commented on models/src/anemoi/models/models/diffusion_encoder_processor_decoder.py:
```diff
@@ -0,0 +1,631 @@
+                grid_shard_shapes = [shape[-2] for shape in shard_shapes]
+                x = shard_tensor(x, -2, shard_shapes, model_comm_group)
+
+            sigma_max = kwargs.get("sigma_max", self.sigma_max)
+            sigma_min = kwargs.get("sigma_min", self.sigma_min)
+            rho = kwargs.get("rho", 7.0)
+            num_steps = kwargs.get("num_steps", 20)
+            sampler = kwargs.get("sampler", "heun")
+            schedule_type = kwargs.get("schedule_type", "karras")
```
Is there a reason for not just adding these to the function signature? Right now these are a bit hidden and easy to miss. We could just add them as arguments with default values?

ssmmnn11 replied:
Changed that now. Can you have another look?

ssmmnn11 commented on models/src/anemoi/models/models/diffusion_encoder_processor_decoder.py:
```diff
@@ -0,0 +1,808 @@
+
+        self.noise_embedder = SinusoidalEmbeddings(num_channels=self.noise_channels, max_period=1000)
+
+        self.noise_cond_mlp = nn.Sequential()
+        self.noise_cond_mlp.add_module("linear1_no_gradscaling", nn.Linear(self.noise_channels, self.noise_channels))
+        self.noise_cond_mlp.add_module("activation", nn.SiLU())
+        self.noise_cond_mlp.add_module("linear2_no_gradscaling", nn.Linear(self.noise_channels, self.noise_cond_dim))
```
I think that is not possible. The names are accessed in the strategy.

https://github.com/ecmwf/anemoi-core/blob/93a6c6b95e5be12b229e4df9160b5dcda428fc01/training/src/anemoi/training/distributed/strategy.py#L49

ssmmnn11 commented on training/docs/user-guide/diffusion-set-up.rst:
```diff
@@ -0,0 +1,195 @@
+-  Forecaster: :class:`GraphDiffusionTendForecaster`
+-  Requires `statistics_tendencies` for normalization
+-  Use single-step rollout (`rollout.max: 1`)
+
```
I should remove that actually

ssmmnn11 commented on models/src/anemoi/models/models/diffusion_encoder_processor_decoder.py:
```diff
@@ -0,0 +1,808 @@
+        x_out = einops.rearrange(x_out, "(bse n) f -> bse n f", bse=bse)
+        x_out = einops.rearrange(x_out, "(bs e) n f -> bs e n f", bs=batch_size).to(dtype=dtype)
+
+        return x_out
```
I think i fixed it

ssmmnn11 commented on models/src/anemoi/models/schemas/models.py:
```diff
@@ -189,6 +222,8 @@ class BaseModelSchema(PydanticBaseModel):
     "GNN decoder schema."
+    diffusion: Optional[DiffusionSchema] = None
+    "Diffusion configuration for diffusion models"
 
```
removed

ssmmnn11 commented on training/src/anemoi/training/distributed/strategy.py:
I needed to make a change and thought it is better to refactor at the same time

mchantry commented on models/src/anemoi/models/distributed/transformer.py:
```diff
@@ -74,6 +74,7 @@ def _seqalltoall(input_: Tensor, shapes: list, group: Optional[ProcessGroup] = N
     # get input format
     input_format = get_memory_format(input_)
 
+    # SL TODO: repair for non sym shapes
```
Do you intend to fix in this PR?

ssmmnn11 commented:
waiting for [basemodule](https://github.com/ecmwf/anemoi-core/pull/399) to be merged

mchantry commented on training/src/anemoi/training/config/model/transformer.yaml:
```diff
@@ -24,7 +24,7 @@ layer_kernels:
   _target_: anemoi.models.layers.processor.TransformerProcessor
   num_layers: 16
-  num_chunks: 2
+  num_chunks: 4
```
Is this a better choice for most setups?

anaprietonem commented on models/src/anemoi/models/layers/diffusion.py:
```diff
@@ -0,0 +1,38 @@
+# (C) Copyright 2024 Anemoi contributors.
```
new files should use the updated header for this year ``Copyright 2025 Anemoi contributors.``

anaprietonem commented on models/src/anemoi/models/models/diffusion_encoder_processor_decoder.py:
```diff
@@ -0,0 +1,809 @@
+# (C) Copyright 2024 Anemoi contributors.
```
same comment as before - we could update to use ``Copyright 2025 Anemoi contributors.``

anaprietonem commented on models/src/anemoi/models/preprocessing/__init__.py:
```diff
@@ -107,7 +107,7 @@ def _invert_key_value_list(method_config: dict[str, list[str]]) -> dict[str, str
         }
 
-    def forward(self, x, in_place: bool = True, inverse: bool = False) -> Tensor:
+    def forward(self, x, in_place: bool = True, inverse: bool = False, **kwargs) -> Tensor:
```
Why do we need this?

anaprietonem commented on training/docs/user-guide/diffusion-set-up.rst:
```diff
@@ -0,0 +1,198 @@
+##########################
+ Diffusion-based training
```
Nice docs!

anaprietonem commented on training/src/anemoi/training/train/train.py:
```diff
@@ -207,7 +207,7 @@ def model(self) -> pl.LightningModule:
             "metadata": self.metadata,
             "statistics": self.datamodule.statistics,
-            "statistics_tendencies": self.datamodule.statistics_tendencies,
+            "statistics_tendencies": self.datamodule.statistics_tendencies,  # should depend on config
```
What do you mean with `# should depend on config`?

anaprietonem replied:
I bet we are passing them because of the tendencies scalers, guess in the future we could construct that kwargs settings based on the config. (so if the config has the Tendency Scalar then we add them). Ultimately the `create_scalars` function accept  `**kwargs` too

@HCookie

ssmmnn11 commented on models/src/anemoi/models/interface/__init__.py:
```diff
@@ -90,9 +90,24 @@ def _build_model(self) -> None:
         self.post_processors = Processors(processors, inverse=True)
 
+        # If tendencies statistics are provided, instantiate the tendencies processors
+        if self.statistics_tendencies is not None:
```
What my thinking was here: only the tendency diffusion model will use these - all model will create them when tendency statistics are available in the dataset. that is not ideal yet.

they are then used e.g. here:

https://github.com/ecmwf/anemoi-core/blob/e98eb2fd64d08785b43fec57945266f17f78bfe0/training/src/anemoi/training/train/tasks/diffusionforecaster.py#L415

https://github.com/ecmwf/anemoi-core/blob/e98eb2fd64d08785b43fec57945266f17f78bfe0/models/src/anemoi/models/models/diffusion_encoder_processor_decoder.py#L595

the aim is to have one set of preprocessors for the tendencies and one for the states. It is true that we can probably have pre  / post processors that won't make sense for tendencies, but I don't know how we can exclude them for now, apart from relying on the user. We could also add extra config options so we can configure tendency post processors differently.

ssmmnn11 commented on training/src/anemoi/training/losses/scalers/loss_weights_mask.py:
```diff
@@ -50,5 +51,8 @@ def get_delayed_scaling_values(self, model: AnemoiModelInterface) -> np.ndarray:
             if hasattr(pre_processor, "loss_mask_training"):
                 loss_weights_mask = loss_weights_mask * pre_processor.loss_mask_training.cpu().numpy()
-
+        if hasattr(model, "pre_processors_tendencies"):
```
yes, these are created in the interface of the model. one issue that I realise now is that we will always update the loss mask for the tendency pre / post processors, even if we don't use them (non tendency diffusion models).

ssmmnn11 commented on training/src/anemoi/training/config/hierarchical.yaml:
```diff
@@ -24,3 +24,5 @@ model:
   enable_hierarchical_level_processing: True
   level_process_num_layers: 2
+  processor:
+    num_chunks: 2
```
we have a chunk parameter for the processor, and we require the number of processor layers to be a multiple of the chunks. I changed the default in this pull request, because we have 2 and for 16 processor layers we usually have, 8 in one go are a bit much (only relevant for the GNN). Here we only have two layers, we could also set num_chunks to 1.

anaprietonem replied:
Seems fine, also guess this would be mostly 'more performant' but should not affect results. @icedoom888 since you use the hierarchical graphs it might be good for you to have visibility but again this seems a reasonable change to me.

ssmmnn11 commented on training/src/anemoi/training/config/training/default.yaml:
```diff
@@ -84,7 +84,7 @@ validation_metrics:
     # - 'loss_weights_mask': Giving imputed NaNs a zero weight in the loss function
     # Use the `scale_validation_metrics` section to variable scale.
-    scalers: []
+    scalers: ['node_weights']
```
I think so. I don't think it makes sense to run without node_weights. But I have not checked the code. Do we get a sum or an average without the node_weights scaler?

anaprietonem replied:
default reduce function for the loss seems to be average https://github.com/ecmwf/anemoi-core/blob/main/training/src/anemoi/training/losses/base.py.

ssmmnn11 replied:
Very good. Though without node weights we don’t form the correct average -> account for different grid spacings

anaprietonem commented on models/src/anemoi/models/samplers/diffusion_samplers.py:
```diff
@@ -0,0 +1,255 @@
+]
+
+
+def get_noise_schedule(
```
Have we tested this in anemoi-inference, do we need changes in the runner too to access the sampler or how will it work?

anaprietonem commented on training/src/anemoi/training/train/tasks/diffusionforecaster.py:
```diff
@@ -0,0 +1,454 @@
+        torch.Tensor
+            Computed loss with noise weighting applied
+        """
+        fact = weights**0.5  # factor for mse loss
```
what is this factor?

HCookie replied:
Potentially

ssmmnn11 replied:
this is linked to the edm diffusion training -> noise level, so with batch size == 2, 0 could have fact x, and 1 fact y. I don't have a super strong opinion, but I think it would make it more complicated weights are computed together with the noise level - also to have the checks in place that for the diffusion model this scaler is required etc.

anaprietonem replied:
Not sure if I get your comment above Simon - do you mean that with bs=2, the samples in the batch could have different value for that factor ? (I think the default config (diffusion.yaml) has a batch size of 2 so I guess I don't understand then why is not like that already or if we should have some asserts/ or at least log this value somewhere in the terminal?)

ssmmnn11 replied:
the factor (weights) differs for each training sample. we would need to replicate this with the scalers. but then maybe it is better to have a diffusion loss instead.

ssmmnn11 commented on training/src/anemoi/training/train/tasks/base.py:
```diff
@@ -356,25 +375,120 @@ def compute_loss_metrics(
+        y_pred: torch.Tensor,
+        y: torch.Tensor,
+        grid_shard_slice: slice | None = None,
+        **_kwargs,
```
I added it because in diffusionforecaster.py we have _compute_loss with the extra weights parameter.

anaprietonem commented on models/src/anemoi/models/models/diffusion_encoder_processor_decoder.py:
```diff
@@ -0,0 +1,809 @@
+        out = einops.rearrange(out, "batch ensemble grid vars -> (batch ensemble grid) vars")
+        return out
+
+    def _generate_noise_conditioning(self, sigma: torch.Tensor, edge_conditioning: bool = False) -> torch.Tensor:
```
How easy do you think would be to adapt this in the future if we have say many encoder or a different graph than one with data-hidden-data?

ssmmnn11 commented on models/src/anemoi/models/samplers/diffusion_samplers.py:
```diff
@@ -0,0 +1,255 @@
+    torch.Tensor
+        Noise schedule with shape (num_steps + 1,)
+    """
+    if schedule_type == "karras":
```
In diffusion training we have:

a) noise scheduler during training 
b) noise schedule during inference / sampling
c) sampler

we need the option to chose b) and c) freely during inference / sampling

a) could be instantiated during training, but we need a plan how to handle inference, which is currently does not have this kind of configurability.

anaprietonem replied:
@gmertes pinging you here for visibility. Might be worth discussing this in the future, since we have touched a few times about config options for inference

anaprietonem commented on training/src/anemoi/training/train/tasks/diffusionforecaster.py:
```diff
@@ -0,0 +1,454 @@
+                # metrics calculation and plotting expects normalised states
+                y = batch[:, self.multi_step + rollout_step, ..., self.data_indices.data.output.full]
+
+                # re-construct target state with noised tendency
```
can this be removed or is that for future work?

anaprietonem commented on models/src/anemoi/models/models/diffusion_encoder_processor_decoder.py:
```diff
@@ -0,0 +1,809 @@
+        x_out = einops.rearrange(x_out, "(bse n) f -> bse n f", bse=bse)
+        x_out = einops.rearrange(x_out, "(bs e) n f -> bs e n f", bs=batch_size).to(dtype=dtype)
+
+        for bounding in self.boundings:
```
In the pydantic schemas there is a check that `Diffusion models do not support bounding layers. ` so do we need this here?

ssmmnn11 replied:
True. I left it in because I think we can (and should?) explore how bounding could come in for the diffusion model. Are you in favour of removing it?

anaprietonem replied:
Yes I was curious why bounding couldn't be used. But until then I'd be in favour of removing as otherwise I think this can be confusing and prone to errors (ie - if people has the config_validation to False then pydantic won't catch that bounding is defined)

anaprietonem commented on training/src/anemoi/training/config/model/graphtransformer_diffusion.yaml:
```diff
@@ -0,0 +1,91 @@
+model:
+  _target_: anemoi.models.models.AnemoiDiffusionModelEncProcDec
+  # Diffusion noise parameters
+  diffusion:
```
Out of curiosity but why did you go for including this inside this block? Do we think those could just specific for the diffusion model or in the future we could also have for the `anemoi.models.models.AnemoiModelEncProcDec` a set of configs entries, so then this could be more general?

ssmmnn11 replied:
I don't remember. might have been related to the fact that we only pass the model part in the tests? 

You think we might have a noise block for the base model as well?

anaprietonem commented on models/src/anemoi/models/schemas/models.py:
```diff
@@ -44,13 +47,23 @@ class DefinedModels(str, Enum):
     "Model object defined in anemoi.models.model."
     convert_: str = Field("all", alias="_convert_")
     "The target's parameters to convert to primitive containers. Other parameters will use OmegaConf. Default to all."
+    diffusion: Optional[DiffusionSchema] = Field(default=None)
```
I have been testing this and I believe as it is currently the diffusion.yaml doesn't model matches the `BaseModelSchema` class and hence the `validate_no_bounding_for_diffusion` doesn't work. (that's why I was asking the motivation for making this a config inside diffusion). Since we don't know yet if in the future we will need a noise block  the base/deterministic models I think it would be better to have:

```

class DiffusionNoiseSchema(BaseModel):
    rho: PositiveFloat = Field(default=7.0, examples=[7.0])
    "Time discretization parameter for Karras schedule"
    sigma_max: PositiveFloat = Field(default=100.0, examples=[100.0])
    "Maximum noise level"
    sigma_min: PositiveFloat = Field(default=0.02, examples=[0.02])
    "Minimum noise level"
    sigma_data: PositiveFloat = Field(default=1.0, examples=[1.0])
    "Data scaling parameter"
    noise_channels: PositiveInt = Field(default=32, examples=[32])
    "Number of channels for noise embedding"
    noise_cond_dim: PositiveInt = Field(default=16, examples=[16])
    "Dimension of noise conditioning"

class DiffusionSchema(BaseModel):
    noise: DiffusionNoiseSchema = Field(default_factory=DiffusionNoiseSchema)
    "Noise configuration for diffusion models"

class DiffusionModel(Model):
    diffusion: DiffusionSchema = Field(default=None)
    "Diffusion configuration for diffusion models"

class DiffusionModelSchema(BaseModelSchema):
    model: DiffusionModel = Field(default_factory=DiffusionModel)
    "Diffusion Model schema"

    @model_validator(mode="after")
    def validate_no_bounding_for_diffusion(self) -> "DiffusionModelSchema":
        if self.bounding:
            msg= (
                "Diffusion models do not support bounding layers. "
                f"Found {len(self.bounding)} bounding configuration(s). "
                "Please remove all bounding configurations for diffusion models."
            )
            raise ValueError(msg)
        return self
        
  ```
Note also that the DiffusionSchema and DifussionNoiseSchema should be `BaseSchema` not the Pydantic one, that's to make sure to check the grammar spelling and that there are no extra/missing keys

anaprietonem commented on models/src/anemoi/models/models/diffusion_encoder_processor_decoder.py:
```diff
@@ -0,0 +1,809 @@
+    def _before_sampling(
+        self,
+        batch: torch.Tensor,
+        pre_processors: nn.Module,
```
the `pre_procesors` are not really used as part of this function right?

anaprietonem commented on models/src/anemoi/models/models/diffusion_encoder_processor_decoder.py:
```diff
@@ -0,0 +1,809 @@
+        model_comm_group: Optional[ProcessGroup] = None,
+        grid_shard_shapes: Optional[list] = None,
+        gather_out: bool = True,
+        pre_processors_tendencies: Optional[nn.Module] = None,
```
`pre_processors_tendencies` can be removed?

icedoom888 reviewed (changes requested):
Thank you for the amazing work @ssmmnn11! 

I have some doubts in the move of `AnemoiModelInterface ` functionality to the model itself. You can see more in the comments.

Also I would refactor some of the code and add more configurable configs to avoid passing redundant and unused parameters throughout the code!

ssmmnn11 commented on models/src/anemoi/models/models/encoder_processor_decoder.py:
```diff
@@ -393,3 +400,68 @@ def forward(
 
         return x_out
+
+    def predict_step(
```
True. The issue is that we currently assume in anemoi inference (and here) that the predict step returns a predicted state. The diffusion model relies on a sampler which calls the model many times, so the predict step becomes more complicated and some of the things we have currently in the predict step do not apply (e.g. when to gather etc.).I agree that we should rethink the need of the interface, but I would do further changes in a follow up PR. This one is already complicated enough and I think refactoring the parts with the diffusion model merges in will be clearer.

anaprietonem replied:
To me the model interface acts as a nice wrapper to use any type of anemoi-model in anemoi-inference. I think it's fair point Alberto, that by moving the predict_step to the tasks this is then likely something we could remove. I think in any case that could be addressed in a future PR.

ssmmnn11 commented on training/src/anemoi/training/config/model/graphtransformer_diffusion.yaml:
```diff
@@ -0,0 +1,91 @@
+    _target_: anemoi.models.layers.normalization.ConditionalLayerNorm
+    normalized_shape: ${model.num_channels}
+    condition_shape: 16
+    w_one_bias_zero_init: True
```
This is the conditional layernorm. But instead of the model embeddings we pass the noise conditioning as inputs to the scale and bias. hence, embeddings will be scaled differently according to the noise level. Sorry, was that the question?

ssmmnn11 replied:
if w_one_bias_zero_init:
            nn.init.zeros_(self.scale.weight)
            nn.init.zeros_(self.scale.bias)
            nn.init.zeros_(self.bias.weight)
            nn.init.zeros_(self.bias.bias)
            
we can remove the _one_ in the future.

ssmmnn11 commented on models/src/anemoi/models/models/diffusion_encoder_processor_decoder.py:
```diff
@@ -0,0 +1,809 @@
+        shape = (batch_size, ensemble_size, grid_size, self.num_output_channels)
+        y_init = torch.randn(shape, device=x.device, dtype=sigmas.dtype) * sigmas[0]
+
+        if sampler == "heun":
```
The diffusion model has two “phases”, training, where we chose a noise schedule - and sampling / inference, where we can chose a sampler and a noise schedule. We want flexibility to chose different noise schedules / samplers at inference time. What you propose would bake that into the model if I understand correctly?

anaprietonem replied:
Had a chat about this with @gmertes, in inference Hydra is not supported (and there is no plan to do so in the short term) so instantiating objects as part of the predict_step would lead to an error. Keeping those as string would work okey but makes the code less clean and the configuration interface for training messier (since there is mix of instantiation and strings). So the options could be:
1. Leave it as it but add a layer to convert from string to instantiate - so at least we can refactor the sampler and noise schedulers to use instantiate and we have a consistent way to load those classes both at inference and training time
2. Consider if we can move the samplers and noise schedulers to be instantiated during the 'build_model'. As it is right now the samplers are generated within the `sample` method

ssmmnn11 replied:
new implementation.

ssmmnn11 commented on models/src/anemoi/models/models/diffusion_encoder_processor_decoder.py:
```diff
@@ -0,0 +1,809 @@
+            raise ValueError(f"Unknown sampler: {sampler}")
+
+
+class AnemoiDiffusionTendModelEncProcDec(AnemoiDiffusionModelEncProcDec):
```
I think good idea for a follow up re-factor. But the task always has to fit the forecaster.

anaprietonem replied:
I like the idea of abstracting this in the future, so using tendencies could be applied to other models aside from diffusion!

icedoom888 commented on models/src/anemoi/models/models/diffusion_encoder_processor_decoder.py:
```diff
@@ -0,0 +1,809 @@
+
+        return x_out
+
+    def fwd_with_preconditioning(
```
I know but this function makes the whole function stack extremely hard to follow for people that didn't code this in the first place..
It is passed to the sampler and is called within it and is called directly in training by the training task, it kinda distupts the usual flow. 
I would find another way of setting this up. Maybe my other suggestion of making the sampler classes and inlcuding them to the state of the model might help with this refactor.

ssmmnn11 commented on models/src/anemoi/models/models/diffusion_encoder_processor_decoder.py:
```diff
@@ -0,0 +1,809 @@
+            truncation_data=truncation_data,
+        )
+
+        self.get_noise_schedule = diffusion_samplers.get_noise_schedule
```
Please see https://github.com/ecmwf/anemoi-core/pull/401#discussion_r2262218304

ssmmnn11 commented:
> > > Thank you for the amazing work @ssmmnn11!
> > > I have some doubts in the move of `AnemoiModelInterface ` functionality to the model itself. You can see more in the comments.
> > > Also I would refactor some of the code and add more configurable configs to avoid passing redundant and unused parameters throughout the code!
> > 
> > 
> > Thank you @icedoom888!
> > I added explanation why I moved the predict step into the model, pls see replies. Are all the configs you mentioned in the comments?
> 
> Hey @ssmmnn11, just saw your replies on some of my comments ;) I am still not 100% convinced that moving the interface logic to the model is the cleanest way to do this, I would leave this up to discussion with the rest of the ECMWF team..
> 
> Regarding noise scheduler and sampler: I propose you initialise them as follows, directly in the init function of the model.
> 
> ```
> self.noise_schedule = instantiate(diffusion_config.noise_schedule)
> self.sampler = instantiate(diffusion_config.sampler ) 
> ```
> 
> This requires two things:
> 
> 1. Modify samplers and schedulers to be classes and not functions.
> 2. a change in the configs where you can set:
>    `diffusion: sampler: _target_: anemoi.models.samplers.EdmHeunSampler S_churn: __ ... noise_schedule: _target_: ...`
>    This would be more configurable and in line with the rest of anemoi codebase.
>    Also please refer to all other comment si made thorughout the code as some of them remain un-addressed :)

Hello @icedoom888,

"Also please refer to all other comment si made thorughout the code as some of them remain un-addressed :)"

I am going through it bit by bit. I also apreciate that diffusion is a bit more complicated than what people are used to and will try my best to explain :-)

Regarding the configs, (I also replied above):

Could you give an example workflow how to set-up sampler and scheduler when we have a checkpoint and run experiments with anemoi-inference with different samplers? I might not be up-to-date with the latest capabilities. In the current form it would go via kwargs in the predict step (to be set in the inference yaml file).

ssmmnn11 commented:
> Hey @ssmmnn11, just saw your replies on some of my comments ;) I am still not 100% convinced that moving the interface logic to the model is the cleanest way to do this, I would leave this up to discussion with the rest of the ECMWF team..
> 

I agree that it could be more elegant, and something that should be refactored in the future. In my opinion it is more important though to get diffusion capabilities into Anemoi, and I don't see how this now affects negatively any experimentation and development. If it does we should avoid it though - could you give an example?

anaprietonem commented on training/src/anemoi/training/losses/scalers/base_scaler.py:
```diff
@@ -91,8 +91,8 @@ class BaseDelayedScaler(BaseScaler):
 
-    def get_delayed_scaling(self, model: AnemoiModelInterface) -> SCALER_DTYPE:
-        scaler_values = self.get_delayed_scaling_values(model)
+    def get_delayed_scaling(self, processors: list[BasePreprocessor]) -> SCALER_DTYPE:
```
I think this can be resolved since @sahahner already looked at it, right?

anaprietonem commented on training/src/anemoi/training/train/tasks/base.py:
```diff
@@ -302,10 +303,22 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
-            self.scalers[name] = scaler_builder.get_delayed_scaling(model=self.model)
+            # Collect processors from the model
+            processors = []
+            processors.extend(self.model.pre_processors.processors.values())
```
This logic is actually updated in main @icedoom888 https://github.com/ecmwf/anemoi-core/blob/main/training/src/anemoi/training/train/tasks/base.py so this will be updated when the branch is updated

ssmmnn11 commented on models/src/anemoi/models/models/diffusion_encoder_processor_decoder.py:
```diff
@@ -0,0 +1,809 @@
+        )
+        c_hidden = self._make_noise_emb(noise_cond, repeat=self.node_attributes.num_nodes[self._graph_name_hidden])
+
+        if edge_conditioning:
```
That is correct. I would like to leave this in as an example how edge conditioning could be implemented when someone is interested in that / an architecture requires it.

ssmmnn11 replied:
I would like to leave it in as an example for people to follow up on if they want to support edge conditioning. I did the similar things when I implemented the GNN/GT/Transformer and ENS architecture  and I think it worked well :-)

anaprietonem commented on models/src/anemoi/models/models/diffusion_encoder_processor_decoder.py:
```diff
@@ -0,0 +1,809 @@
+        pred = self(
+            x, (c_in * y_noised), c_noise, model_comm_group=model_comm_group, grid_shard_shapes=grid_shard_shapes
+        )  # calls forward ...
+        D_x = c_skip * y_noised + c_out * pred
```
I think adding as a comment or docstring of the function sounds good

ssmmnn11 replied:
I refer to EDM now. We can also add links to publications consistenly across the code base?

anaprietonem commented on training/src/anemoi/training/config/model/graphtransformer_diffusion.yaml:
```diff
@@ -0,0 +1,107 @@
+    sigma_max: 100.0
+    sigma_min: 0.02
+    rho: 7.0
+    noise_embedder:
```
I think this new interface for the config looks much better and addresses the points raised by @icedoom888 , thanks @ssmmnn11 !

ssmmnn11 commented on models/src/anemoi/models/schemas/models.py:
```diff
@@ -209,6 +242,23 @@ class NoiseInjectorSchema(BaseModel):
 class EnsModelSchema(BaseModelSchema):
     noise_injector: NoiseInjectorSchema = Field(default_factory=list)
     "Settings related to custom kernels for encoder processor and decoder blocks"
+    diffusion: Optional[DiffusionSchema] = None
```
on the todo list :-)

ssmmnn11 commented on training/src/anemoi/training/losses/scalers/loss_weights_mask.py:
```diff
@@ -32,24 +32,31 @@ def __init__(self, norm: str | None = None, **kwargs) -> None:
+
+        processors = [model.pre_processors]
+        if self.use_processors_tendencies and hasattr(model, "pre_processors_tendencies"):
+            processors.extend(model.pre_processors_tendencies)
```
A combination of both are applied, so I don't want to duplicate the loss mask computation for both. I think we will have to think about how the processors can be made more suitable for this kind of application.

ssmmnn11 replied:
can I resolve?

sahahner replied:
I am not sure this should be the final logic here, but for the nan mask scaler, the mask is just a 0-1-matrix, so multiplying it twice is fine. Is it the same set of pre/postprocessors that are applied twice?

ssmmnn11 replied:
Yes, it would just be multiplying with the same thing twice, which in my opinion does not make sense. Let's do a follow up PR if there are issue with this logic.

ssmmnn11 replied:
for example here:

https://github.com/ecmwf/anemoi-core/blob/e6abc8ba6c7215834e7373ecd9458bcd99b81451/training/src/anemoi/training/train/tasks/diffusionforecaster.py#L393

ssmmnn11 commented on models/src/anemoi/models/models/diffusion_encoder_processor_decoder.py:
```diff
@@ -0,0 +1,809 @@
+
+        self.get_noise_schedule = diffusion_samplers.get_noise_schedule
+
+        self.noise_embedder = SinusoidalEmbeddings(num_channels=self.noise_channels, max_period=1000)
```
done

ssmmnn11 commented on models/src/anemoi/models/models/diffusion_encoder_processor_decoder.py:
```diff
@@ -0,0 +1,809 @@
+        num_steps: int = 20,
+        sampler: str = "heun",
+        schedule_type: str = "karras",
+        S_churn: float = 0.0,
```
see new implementation pls

VeraChristina commented on training/tests/integration/test_training_cycle.py:
```diff
@@ -182,3 +182,16 @@ def test_config_validation_interpolator(interpolator_config: tuple[DictConfig, s
+
+
+@skip_if_offline
+@pytest.mark.longtests
```
Yes, exactly!
``` @pytest.mark.slow ```

anaprietonem reviewed (approved):
Comments and suggestion have been addressed so we consider this is now good to be merged. @ssmmnn11 great work!! 🚀 and thanks for all the efforts. 

@icedoom888 we consider the suggestions you raised are implemented, but if upon your return from holidays there are still things you want to discuss please reach out. 

thanks to @sahahner, @HCookie and @gmertes for insights and review!
